pubtools/_content_gateway/push_base.py

```diff
@@ -480,33 +534,140 @@ def process_file(self, file_item):
+                    self.completed_operations.remove(record)
+
+            LOG.info("File record deleted! \n")
             del self.file_mapping[(product_name, product_code, version_name, download_url)]
         else:
-            raise CGWError("Cannot delete file: product version: '%s' product code: '%s' product: '%s', id is not set" %
-                           (version_name, product_code, product_name))
+            raise CGWError(
+                "Cannot update/delete file: product version: '%s' product code: '%s' product: '%s', id is not set" %
+                (version_name, product_code, product_name))
         return file_id
+
+    def make_visible(self):
+        """
+        - Update the invisible attribute to False
+          once all the CGW operation completed successfully.
+
+        - This is needed to make the version and file records visible.
+        - The `invisible` attribute changes on the following conditions:
+
+            - `invisible=True` :- sets if user externally specify the `invisible=True`
+            - `invisible=False` :- sets either if user externally specify the `invisible=False`
+              or user doesn't specify any value for this `invisible` attribute.
+
+        """
+
+        LOG.debug("Updating the invisible attribute of the created records")
+        LOG.debug("Enabling the created records if required \n")
+
+        self.completed_operations.reverse()
+        for item in self.completed_operations:
+            if item.get('type') == 'product_version':
+                if item.get("state") == "create":
+                    item["metadata"]["id"] = item["version_id"]
+                    item["metadata"]['invisible'] = False if not item["metadata"].get('invisible') else True
+                    self.cgw_client.update_version(item["metadata"]["productId"], item["metadata"])
+
+            elif item.get('type') == 'file':
+                if item.get("state") == "create":
+                    item["metadata"]["id"] = item["file_id"]
+                    item["metadata"]['invisible'] = False if not item["metadata"].get('invisible') else True
+                    self.cgw_client.update_file(
+                        item["product_id"],
+                        item["metadata"]["productVersionId"],
+                        item["metadata"]
+                    )
+
+    def rollback_cgw_operation(self):
+        """
+        Undo the partial completed CGW operation.
+            - For the create operations it will delete the created data from CGW
+            - For the update operation it will revert the changes with old data.
+
+        """
+
+        self.completed_operations.reverse()
+        for item in self.completed_operations:
+            if item.get('type') == 'product':
+                if item.get("state") == "create":
+                    self.cgw_client.delete_product(item["product_id"])
+                    continue
```

Comment:
are these "continues" basically unnecessary?

Reply:
It is not but it went evaluating the next elif statement that I wanted to prevent. But we can remove it as it is preventing us just to evaluate a single line.